In [ ]:
from google.colab import drive
import pandas as pd

# Read the CSV file from drive
drive.mount('/content/drive')
df = pd.read_csv('/content/drive/MyDrive/Airbnb_Data/airnb.csv')
#df2 = pd.read_csv('/content/drive/MyDrive/Airbnb_Data/airnb_desert.csv', encoding='latin1')
#df3 = pd.read_csv('/content/drive/MyDrive/Airbnb_Data/airnb_luxe.csv', encoding='latin1')

display(df.head())
#display(df2.head())
#display(df3.head())

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


,Title,Detail,Date,Price(in dollar),Offer price(in dollar),Review and rating,Number of bed
0,"Chalet in Skykomish, Washington, US",Sky Haus - A-Frame Cabin,Jun 11 - 16,306.00,229.00,4.85 (531),4 beds
1,"Cabin in Hancock, New York, US",The Catskill A-Frame - Mid-Century Modern Cabin,Jun 6 - 11,485.00,170.00,4.77 (146),4 beds
2,"Cabin in West Farmington, Ohio, US",The Triangle: A-Frame Cabin for your city retreat,Jul 9 - 14,119.00,522.00,4.91 (515),4 beds
3,"Home in Blue Ridge, Georgia, US",*Summer Sizzle* 5 Min to Blue Ridge* Pets* Hot...,Jun 11 - 16,192.00,348.00,4.94 (88),5 beds
4,"Treehouse in Grandview, Texas, US",Luxury Treehouse Couples Getaway w/ Peaceful V...,Jun 4 - 9,232.00,196.00,4.99 (222),1 queen bed


In [ ]:
df.info()
df.describe(include="all")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 953 entries, 0 to 952
Data columns (total 7 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Title                   953 non-null    object
 1   Detail                  953 non-null    object
 2   Date                    953 non-null    object
 3   Price(in dollar)        953 non-null    object
 4   Offer price(in dollar)  166 non-null    object
 5   Review and rating       947 non-null    object
 6   Number of bed           953 non-null    object
dtypes: object(7)
memory usage: 52.2+ KB


,Title,Detail,Date,Price(in dollar),Offer price(in dollar),Review and rating,Number of bed
count,953,953,953,953,166,947,953
unique,756,839,204,322,108,754,33
top,"Apartment in Kuala Lumpur, Malaysia",Cozy room 48.5 sq.m @The Astra near Night market,May 1 - 6,111.00,63.00,New,2 beds
freq,7,3,219,13,4,16,221


In [ ]:
df.detail = (
    df.Detail
    .str.strip()
    .str.lower()
    .str.replace(" ", "_", regex=False)
    .str.replace("/", "_", regex=False)
    .str.replace("*", "_", regex=False)
    .str.replace(r"[^a-z0-9_]", "", regex=True)
)
display (df)

/tmp/ipykernel_1565/511648279.py:1: UserWarning: Pandas doesn't allow columns to be created via a new attribute name - see https://pandas.pydata.org/pandas-docs/stable/indexing.html#attribute-access
  df.detail = (


,Title,Detail,Date,Price(in dollar),Offer price(in dollar),Review and rating,Number of bed
0,"Chalet in Skykomish, Washington, US",Sky Haus - A-Frame Cabin,Jun 11 - 16,306.00,229.00,4.85 (531),4 beds
1,"Cabin in Hancock, New York, US",The Catskill A-Frame - Mid-Century Modern Cabin,Jun 6 - 11,485.00,170.00,4.77 (146),4 beds
2,"Cabin in West Farmington, Ohio, US",The Triangle: A-Frame Cabin for your city retreat,Jul 9 - 14,119.00,522.00,4.91 (515),4 beds
3,"Home in Blue Ridge, Georgia, US",*Summer Sizzle* 5 Min to Blue Ridge* Pets* Hot...,Jun 11 - 16,192.00,348.00,4.94 (88),5 beds
4,"Treehouse in Grandview, Texas, US",Luxury Treehouse Couples Getaway w/ Peaceful V...,Jun 4 - 9,232.00,196.00,4.99 (222),1 queen bed
...,...,...,...,...,...,...,...
948,Guest suite in Lake Worth,The Garden Suite at the Historic Chateau Bleu,May 1 - 6,141.00,NaN,4.85 (82),1 king bed
949,Room in Kissimmee,Private Room in Orlando/Kissimmee (2Twins) T1,May 1 - 6,43.00,NaN,4.9 (300),1 king bed
950,Guesthouse in San Antonio,Casa Alexander,May 1 - 6,73.00,NaN,4.92 (282),5 beds
951,Apartment in Englewood,BEACH FRONT Unit w/ Pool Sandy Toe's # 5,May 7 - 12,233.00,NaN,4.96 (318),4 beds


In [ ]:
df["Date"].head(20)

,Date
0,Jun 11 - 16
1,Jun 6 - 11
2,Jul 9 - 14
3,Jun 11 - 16
4,Jun 4 - 9
5,Jun 21 - 26
6,Jun 4 - 9
7,Jun 8 - 15
8,Jul 3 - 8
9,Jun 26 - Jul 1


In [ ]:
import pandas as pd

def title_extractor(df: pd.DataFrame) -> pd.DataFrame:
    temp_df = df.copy()
    temp_df[["Header", "State", "Country"]] = temp_df["Title"].str.split(
        ",", expand=True
    )
    temp_df.drop("Title", inplace=True, axis=1)
    return temp_df

In [ ]:
df_extracted = title_extractor(df)
display(df_extracted.head(30))

,Detail,Date,Price(in dollar),Offer price(in dollar),Review and rating,Number of bed,Header,State,Country
0,Sky Haus - A-Frame Cabin,Jun 11 - 16,306.00,229.00,4.85 (531),4 beds,Chalet in Skykomish,Washington,US
1,The Catskill A-Frame - Mid-Century Modern Cabin,Jun 6 - 11,485.00,170.00,4.77 (146),4 beds,Cabin in Hancock,New York,US
2,The Triangle: A-Frame Cabin for your city retreat,Jul 9 - 14,119.00,522.00,4.91 (515),4 beds,Cabin in West Farmington,Ohio,US
3,*Summer Sizzle* 5 Min to Blue Ridge* Pets* Hot...,Jun 11 - 16,192.00,348.00,4.94 (88),5 beds,Home in Blue Ridge,Georgia,US
4,Luxury Treehouse Couples Getaway w/ Peaceful V...,Jun 4 - 9,232.00,196.00,4.99 (222),1 queen bed,Treehouse in Grandview,Texas,US
5,Casa Tiny near Casa Wabi,Jun 21 - 26,261.00,148.00,4.84 (555),1 double bed,Tiny home in Puerto Escondido,Mexico,None
6,The Beach House. Oceanfront & Sunrise Views!,Jun 4 - 9,607.00,NaN,4.64 (90),2 queen beds,Home in Saco,Maine,US
7,• Cedar Peak • 2 Bedroom Barrier-Free Chalet,Jun 8 - 15,381.00,NaN,5.0 (48),2 beds,Chalet in Grand Étang,Canada,None
8,"Room in apartment, traditional neighborhood",Jul 3 - 8,29.00,NaN,4.84 (58),2 beds,Room in Mexico City,Mexico,None
9,Peacock Tiny House near Las Vegas,Jun 26 - Jul 1,154.00,NaN,4.92 (744),3 beds,Tiny home in Sandy Valley,Nevada,US


In [ ]:
df_extracted.columns = [(column.lower()).replace(" ","_") for column in df_extracted.columns]
display(df_extracted)

,detail,date,price(in_dollar),offer_price(in_dollar),review_and_rating,number_of_bed,header,state,country
0,Sky Haus - A-Frame Cabin,Jun 11 - 16,306.00,229.00,4.85 (531),4 beds,Chalet in Skykomish,Washington,US
1,The Catskill A-Frame - Mid-Century Modern Cabin,Jun 6 - 11,485.00,170.00,4.77 (146),4 beds,Cabin in Hancock,New York,US
2,The Triangle: A-Frame Cabin for your city retreat,Jul 9 - 14,119.00,522.00,4.91 (515),4 beds,Cabin in West Farmington,Ohio,US
3,*Summer Sizzle* 5 Min to Blue Ridge* Pets* Hot...,Jun 11 - 16,192.00,348.00,4.94 (88),5 beds,Home in Blue Ridge,Georgia,US
4,Luxury Treehouse Couples Getaway w/ Peaceful V...,Jun 4 - 9,232.00,196.00,4.99 (222),1 queen bed,Treehouse in Grandview,Texas,US
...,...,...,...,...,...,...,...,...,...
948,The Garden Suite at the Historic Chateau Bleu,May 1 - 6,141.00,NaN,4.85 (82),1 king bed,Guest suite in Lake Worth,None,None
949,Private Room in Orlando/Kissimmee (2Twins) T1,May 1 - 6,43.00,NaN,4.9 (300),1 king bed,Room in Kissimmee,None,None
950,Casa Alexander,May 1 - 6,73.00,NaN,4.92 (282),5 beds,Guesthouse in San Antonio,None,None
951,BEACH FRONT Unit w/ Pool Sandy Toe's # 5,May 7 - 12,233.00,NaN,4.96 (318),4 beds,Apartment in Englewood,None,None


In [ ]:
df_extracted[['raiting','reviews']] = df_extracted.review_and_rating.str.split(" ", expand=True)
display(df_extracted)

,detail,date,price(in_dollar),offer_price(in_dollar),review_and_rating,number_of_bed,header,state,country,raiting,reviews
0,Sky Haus - A-Frame Cabin,Jun 11 - 16,306.00,229.00,4.85 (531),4 beds,Chalet in Skykomish,Washington,US,4.85,(531)
1,The Catskill A-Frame - Mid-Century Modern Cabin,Jun 6 - 11,485.00,170.00,4.77 (146),4 beds,Cabin in Hancock,New York,US,4.77,(146)
2,The Triangle: A-Frame Cabin for your city retreat,Jul 9 - 14,119.00,522.00,4.91 (515),4 beds,Cabin in West Farmington,Ohio,US,4.91,(515)
3,*Summer Sizzle* 5 Min to Blue Ridge* Pets* Hot...,Jun 11 - 16,192.00,348.00,4.94 (88),5 beds,Home in Blue Ridge,Georgia,US,4.94,(88)
4,Luxury Treehouse Couples Getaway w/ Peaceful V...,Jun 4 - 9,232.00,196.00,4.99 (222),1 queen bed,Treehouse in Grandview,Texas,US,4.99,(222)
...,...,...,...,...,...,...,...,...,...,...,...
948,The Garden Suite at the Historic Chateau Bleu,May 1 - 6,141.00,NaN,4.85 (82),1 king bed,Guest suite in Lake Worth,None,None,4.85,(82)
949,Private Room in Orlando/Kissimmee (2Twins) T1,May 1 - 6,43.00,NaN,4.9 (300),1 king bed,Room in Kissimmee,None,None,4.9,(300)
950,Casa Alexander,May 1 - 6,73.00,NaN,4.92 (282),5 beds,Guesthouse in San Antonio,None,None,4.92,(282)
951,BEACH FRONT Unit w/ Pool Sandy Toe's # 5,May 7 - 12,233.00,NaN,4.96 (318),4 beds,Apartment in Englewood,None,None,4.96,(318)


In [ ]:
df_extracted["reviews"] = df_extracted["reviews"].str.replace(r'[()]', '', regex=True)
display(df_extracted)

,detail,date,price(in_dollar),offer_price(in_dollar),review_and_rating,number_of_bed,header,state,country,raiting,reviews
0,Sky Haus - A-Frame Cabin,Jun 11 - 16,306.00,229.00,4.85 (531),4 beds,Chalet in Skykomish,Washington,US,4.85,531
1,The Catskill A-Frame - Mid-Century Modern Cabin,Jun 6 - 11,485.00,170.00,4.77 (146),4 beds,Cabin in Hancock,New York,US,4.77,146
2,The Triangle: A-Frame Cabin for your city retreat,Jul 9 - 14,119.00,522.00,4.91 (515),4 beds,Cabin in West Farmington,Ohio,US,4.91,515
3,*Summer Sizzle* 5 Min to Blue Ridge* Pets* Hot...,Jun 11 - 16,192.00,348.00,4.94 (88),5 beds,Home in Blue Ridge,Georgia,US,4.94,88
4,Luxury Treehouse Couples Getaway w/ Peaceful V...,Jun 4 - 9,232.00,196.00,4.99 (222),1 queen bed,Treehouse in Grandview,Texas,US,4.99,222
...,...,...,...,...,...,...,...,...,...,...,...
948,The Garden Suite at the Historic Chateau Bleu,May 1 - 6,141.00,NaN,4.85 (82),1 king bed,Guest suite in Lake Worth,None,None,4.85,82
949,Private Room in Orlando/Kissimmee (2Twins) T1,May 1 - 6,43.00,NaN,4.9 (300),1 king bed,Room in Kissimmee,None,None,4.9,300
950,Casa Alexander,May 1 - 6,73.00,NaN,4.92 (282),5 beds,Guesthouse in San Antonio,None,None,4.92,282
951,BEACH FRONT Unit w/ Pool Sandy Toe's # 5,May 7 - 12,233.00,NaN,4.96 (318),4 beds,Apartment in Englewood,None,None,4.96,318


In [ ]:
df_extracted.drop(['detail', 'date','number_of_bed','review_and_rating','offer_price(in_dollar)'], axis=1, inplace=True)
display(df_extracted)
#df_extracted.dropna(subset=['detail', 'date','review_and_rating','number_of_bed','raiting_and_reviews'])

,price(in_dollar),header,state,country,raiting,reviews
0,306.00,Chalet in Skykomish,Washington,US,4.85,531
1,485.00,Cabin in Hancock,New York,US,4.77,146
2,119.00,Cabin in West Farmington,Ohio,US,4.91,515
3,192.00,Home in Blue Ridge,Georgia,US,4.94,88
4,232.00,Treehouse in Grandview,Texas,US,4.99,222
...,...,...,...,...,...,...
948,141.00,Guest suite in Lake Worth,None,None,4.85,82
949,43.00,Room in Kissimmee,None,None,4.9,300
950,73.00,Guesthouse in San Antonio,None,None,4.92,282
951,233.00,Apartment in Englewood,None,None,4.96,318


In [ ]:
df_extracted["price(in_dollar)"] = pd.to_numeric(
    df_extracted["price(in_dollar)"],
    errors="coerce"
)

df_extracted["raiting"] = pd.to_numeric(
    df_extracted["raiting"],
    errors="coerce"
)

df_extracted["reviews"] = pd.to_numeric(
    df_extracted["reviews"],
    errors="coerce"
)

In [ ]:
df_extracted.columns.tolist()

['price(in_dollar)', 'header', 'state', 'country', 'raiting', 'reviews']

In [ ]:
df_extracted["header"].head(20).tolist()

['Chalet in Skykomish',
 'Cabin in Hancock',
 'Cabin in West Farmington',
 'Home in Blue Ridge',
 'Treehouse in Grandview',
 'Tiny home in Puerto Escondido',
 'Home in Saco',
 'Chalet in Grand Étang',
 'Room in Mexico City',
 'Tiny home in Sandy Valley',
 'Guest suite in Huntsville',
 'Cabin in Nelson',
 'Dome in Joshua Tree',
 'Barn in Vero Beach',
 'Tiny home in Bolton',
 'Cabin in Ucluelet',
 'Barn in Essex',
 'Apartment in Bacalar',
 'Treehouse in Travelers Rest',
 'Farm stay in Del Norte County']

In [ ]:
def header_extractor(df: pd.DataFrame) -> pd.DataFrame:
    temp_df = df.copy()

    # Split header into type and location
    split_header = temp_df["header"].str.split(" in ", n=1, expand=True)

    # Give the two resulting columns names
    split_header.columns = ["type", "location"]

    # Add them to the dataframe
    temp_df[["type", "location"]] = split_header

    # Remove original header
    temp_df.drop(columns=["header"], inplace=True)

    return temp_df

In [ ]:
df_location = header_extractor(df_extracted)
display(df_location.head(30))

,price(in_dollar),state,country,raiting,reviews,type,location
0,306.0,Washington,US,4.85,531.0,Chalet,Skykomish
1,485.0,New York,US,4.77,146.0,Cabin,Hancock
2,119.0,Ohio,US,4.91,515.0,Cabin,West Farmington
3,192.0,Georgia,US,4.94,88.0,Home,Blue Ridge
4,232.0,Texas,US,4.99,222.0,Treehouse,Grandview
5,261.0,Mexico,None,4.84,555.0,Tiny home,Puerto Escondido
6,607.0,Maine,US,4.64,90.0,Home,Saco
7,381.0,Canada,None,5.00,48.0,Chalet,Grand Étang
8,29.0,Mexico,None,4.84,58.0,Room,Mexico City
9,154.0,Nevada,US,4.92,744.0,Tiny home,Sandy Valley


In [ ]:
df_location_copy=df_location.copy()

Airbnb data
     ↓
unique locations
     ↓
geocoding
     ↓
latitude + longitude
     ↓
JOIN back to Airbnb data??

In [ ]:
duplicates_total = df_location_copy.duplicated().sum()
print(f"Total of duplicated rows is: {duplicates_total}")
#duplicate_rows = df_location_copy[df_location_copy.duplicated()]
#duplicate_rows.head(15)

Total of duplicated rows is: 96


In [ ]:
df_location_copy.drop_duplicates(inplace=True)

In [ ]:
df_location_copy.info()

<class 'pandas.core.frame.DataFrame'>
Index: 857 entries, 0 to 952
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   price(in_dollar)  856 non-null    float64
 1   state             622 non-null    object 
 2   country           117 non-null    object 
 3   raiting           837 non-null    float64
 4   reviews           837 non-null    float64
 5   type              857 non-null    object 
 6   location          857 non-null    object 
dtypes: float64(3), object(4)
memory usage: 53.6+ KB


In [ ]:
df_location_copy.isna().sum()

,0
price(in_dollar),1
state,235
country,740
raiting,20
reviews,20
type,0
location,0


In [ ]:
df_location_copy[df_location_copy["country"].isna()]

,price(in_dollar),state,country,raiting,reviews,type,location
5,261.0,Mexico,None,4.84,555.0,Tiny home,Puerto Escondido
7,381.0,Canada,None,5.00,48.0,Chalet,Grand Étang
8,29.0,Mexico,None,4.84,58.0,Room,Mexico City
10,207.0,Canada,None,4.97,265.0,Guest suite,Huntsville
11,162.0,Canada,None,4.99,226.0,Cabin,Nelson
...,...,...,...,...,...,...,...
948,141.0,None,None,4.85,82.0,Guest suite,Lake Worth
949,43.0,None,None,4.90,300.0,Room,Kissimmee
950,73.0,None,None,4.92,282.0,Guesthouse,San Antonio
951,233.0,None,None,4.96,318.0,Apartment,Englewood


In [ ]:
#Transform objects into string for rolls "country","state" and "location"


**IMPORTANTE:** IN THE PAST CELL I FOUND OUT THAT THERE ARE ACTUALLY 740 NONE(NULL) VALUES IN THE COLLUMN COUNTRY
---



So it is really necessary to fill those values, using the information from the states, but to address the problem in a doable way, I will initially work with a subdata frame for the US

In [ ]:
us_states = [
    "Alabama", "Alaska", "Arizona", "Arkansas", "California",
    "Colorado", "Connecticut", "Delaware", "Florida", "Georgia",
    "Hawaii", "Idaho", "Illinois", "Indiana", "Iowa", "Kansas",
    "Kentucky", "Louisiana", "Maine", "Maryland", "Massachusetts",
    "Michigan", "Minnesota", "Mississippi", "Missouri", "Montana",
    "Nebraska", "Nevada", "New Hampshire", "New Jersey", "New Mexico",
    "New York", "North Carolina", "North Dakota", "Ohio", "Oklahoma",
    "Oregon", "Pennsylvania", "Rhode Island", "South Carolina",
    "South Dakota", "Tennessee", "Texas", "Utah", "Vermont",
    "Virginia", "Washington", "West Virginia", "Wisconsin", "Wyoming"
]

In [ ]:
mask = (
    df_location_copy["country"].isna()
    & df_location_copy["state"].isin(us_states)
)

df_location_copy.loc[mask, "country"] = "US"
display(df_location_copy.head(30))

,price(in_dollar),state,country,raiting,reviews,type,location
0,306.0,Washington,US,4.85,531.0,Chalet,Skykomish
1,485.0,New York,US,4.77,146.0,Cabin,Hancock
2,119.0,Ohio,US,4.91,515.0,Cabin,West Farmington
3,192.0,Georgia,US,4.94,88.0,Home,Blue Ridge
4,232.0,Texas,US,4.99,222.0,Treehouse,Grandview
5,261.0,Mexico,None,4.84,555.0,Tiny home,Puerto Escondido
6,607.0,Maine,US,4.64,90.0,Home,Saco
7,381.0,Canada,None,5.00,48.0,Chalet,Grand Étang
8,29.0,Mexico,None,4.84,58.0,Room,Mexico City
9,154.0,Nevada,US,4.92,744.0,Tiny home,Sandy Valley


In [ ]:
df_location_copy.dropna(subset=["state"])
display(df_location_copy.head(30))

,price(in_dollar),state,country,raiting,reviews,type,location
0,306.0,Washington,US,4.85,531.0,Chalet,Skykomish
1,485.0,New York,US,4.77,146.0,Cabin,Hancock
2,119.0,Ohio,US,4.91,515.0,Cabin,West Farmington
3,192.0,Georgia,US,4.94,88.0,Home,Blue Ridge
4,232.0,Texas,US,4.99,222.0,Treehouse,Grandview
5,261.0,Mexico,None,4.84,555.0,Tiny home,Puerto Escondido
6,607.0,Maine,US,4.64,90.0,Home,Saco
7,381.0,Canada,None,5.00,48.0,Chalet,Grand Étang
8,29.0,Mexico,None,4.84,58.0,Room,Mexico City
9,154.0,Nevada,US,4.92,744.0,Tiny home,Sandy Valley


In [ ]:
print(df_location_copy["country"].unique())

[' US' None ' Mexico' ' Italy' ' Thailand' ' Indonesia' ' Philippines'
 ' Malaysia']


In [ ]:
us_locations = df_location_copy[
    df_location_copy["country"] == " US"
][["location", "state"]]
display(us_locations.head(30))

,location,state
0,Skykomish,Washington
1,Hancock,New York
2,West Farmington,Ohio
3,Blue Ridge,Georgia
4,Grandview,Texas
6,Saco,Maine
9,Sandy Valley,Nevada
12,Joshua Tree,California
13,Vero Beach,Florida
14,Bolton,Vermont


In [ ]:
df_location_us = df_location_copy[df_location_copy["country"] == " US"].drop_duplicates()
display(df_location_us.head(30))
print(df_location_us.shape)

,price(in_dollar),state,country,raiting,reviews,type,location
0,306.0,Washington,US,4.85,531.0,Chalet,Skykomish
1,485.0,New York,US,4.77,146.0,Cabin,Hancock
2,119.0,Ohio,US,4.91,515.0,Cabin,West Farmington
3,192.0,Georgia,US,4.94,88.0,Home,Blue Ridge
4,232.0,Texas,US,4.99,222.0,Treehouse,Grandview
6,607.0,Maine,US,4.64,90.0,Home,Saco
9,154.0,Nevada,US,4.92,744.0,Tiny home,Sandy Valley
12,268.0,California,US,4.94,944.0,Dome,Joshua Tree
13,111.0,Florida,US,4.95,417.0,Barn,Vero Beach
14,174.0,Vermont,US,4.97,372.0,Tiny home,Bolton


(108, 7)


In [ ]:
df_location_us["google_maps"] = (
    "https://www.google.com/maps/search/?api=1&query="
    + df_location_us["location"].fillna("") + ", "
    + df_location_us["state"].fillna("") + ", "
    + df_location_us["country"].fillna("")
)

In [ ]:
print(df_location_us.columns.tolist())

['price(in_dollar)', 'state', 'country', 'raiting', 'reviews', 'type', 'location', 'google_maps']


In [ ]:
display(
    df_location_us[
        ["location", "state", "country", "google_maps"]
    ].head(20)
)

,location,state,country,google_maps
0,Skykomish,Washington,US,https://www.google.com/maps/search/?api=1&quer...
1,Hancock,New York,US,https://www.google.com/maps/search/?api=1&quer...
2,West Farmington,Ohio,US,https://www.google.com/maps/search/?api=1&quer...
3,Blue Ridge,Georgia,US,https://www.google.com/maps/search/?api=1&quer...
4,Grandview,Texas,US,https://www.google.com/maps/search/?api=1&quer...
6,Saco,Maine,US,https://www.google.com/maps/search/?api=1&quer...
9,Sandy Valley,Nevada,US,https://www.google.com/maps/search/?api=1&quer...
12,Joshua Tree,California,US,https://www.google.com/maps/search/?api=1&quer...
13,Vero Beach,Florida,US,https://www.google.com/maps/search/?api=1&quer...
14,Bolton,Vermont,US,https://www.google.com/maps/search/?api=1&quer...


In [ ]:
from IPython.display import display, HTML

display(
    HTML(
        df_location_us[
            ["location", "state", "country", "google_maps"]
        ].head(20).to_html(
            render_links=True,
            escape=False
        )
    )
)

,location,state,country,google_maps
0,Skykomish,Washington,US,"https://www.google.com/maps/search/?api=1&query=Skykomish, Washington, US"
1,Hancock,New York,US,"https://www.google.com/maps/search/?api=1&query=Hancock, New York, US"
2,West Farmington,Ohio,US,"https://www.google.com/maps/search/?api=1&query=West Farmington, Ohio, US"
3,Blue Ridge,Georgia,US,"https://www.google.com/maps/search/?api=1&query=Blue Ridge, Georgia, US"
4,Grandview,Texas,US,"https://www.google.com/maps/search/?api=1&query=Grandview, Texas, US"
6,Saco,Maine,US,"https://www.google.com/maps/search/?api=1&query=Saco, Maine, US"
9,Sandy Valley,Nevada,US,"https://www.google.com/maps/search/?api=1&query=Sandy Valley, Nevada, US"
12,Joshua Tree,California,US,"https://www.google.com/maps/search/?api=1&query=Joshua Tree, California, US"
13,Vero Beach,Florida,US,"https://www.google.com/maps/search/?api=1&query=Vero Beach, Florida, US"
14,Bolton,Vermont,US,"https://www.google.com/maps/search/?api=1&query=Bolton, Vermont, US"


In [ ]:
display(df_location_us.head(10))

,price(in_dollar),state,country,raiting,reviews,type,location,google_maps
0,306.0,Washington,US,4.85,531.0,Chalet,Skykomish,https://www.google.com/maps/search/?api=1&quer...
1,485.0,New York,US,4.77,146.0,Cabin,Hancock,https://www.google.com/maps/search/?api=1&quer...
2,119.0,Ohio,US,4.91,515.0,Cabin,West Farmington,https://www.google.com/maps/search/?api=1&quer...
3,192.0,Georgia,US,4.94,88.0,Home,Blue Ridge,https://www.google.com/maps/search/?api=1&quer...
4,232.0,Texas,US,4.99,222.0,Treehouse,Grandview,https://www.google.com/maps/search/?api=1&quer...
6,607.0,Maine,US,4.64,90.0,Home,Saco,https://www.google.com/maps/search/?api=1&quer...
9,154.0,Nevada,US,4.92,744.0,Tiny home,Sandy Valley,https://www.google.com/maps/search/?api=1&quer...
12,268.0,California,US,4.94,944.0,Dome,Joshua Tree,https://www.google.com/maps/search/?api=1&quer...
13,111.0,Florida,US,4.95,417.0,Barn,Vero Beach,https://www.google.com/maps/search/?api=1&quer...
14,174.0,Vermont,US,4.97,372.0,Tiny home,Bolton,https://www.google.com/maps/search/?api=1&quer...


In [ ]:
#Interactive table with links to Google Maps
from IPython.display import display, HTML

display(
    HTML(
        df_location_us[
            ["type","reviews","raiting","price(in_dollar)","location", "state", "country", "google_maps"]
        ].head(20).to_html(
            render_links=True,
            escape=False
        )
    )
)

,type,reviews,raiting,price(in_dollar),location,state,country,google_maps
0,Chalet,531.0,4.85,306.0,Skykomish,Washington,US,"https://www.google.com/maps/search/?api=1&query=Skykomish, Washington, US"
1,Cabin,146.0,4.77,485.0,Hancock,New York,US,"https://www.google.com/maps/search/?api=1&query=Hancock, New York, US"
2,Cabin,515.0,4.91,119.0,West Farmington,Ohio,US,"https://www.google.com/maps/search/?api=1&query=West Farmington, Ohio, US"
3,Home,88.0,4.94,192.0,Blue Ridge,Georgia,US,"https://www.google.com/maps/search/?api=1&query=Blue Ridge, Georgia, US"
4,Treehouse,222.0,4.99,232.0,Grandview,Texas,US,"https://www.google.com/maps/search/?api=1&query=Grandview, Texas, US"
6,Home,90.0,4.64,607.0,Saco,Maine,US,"https://www.google.com/maps/search/?api=1&query=Saco, Maine, US"
9,Tiny home,744.0,4.92,154.0,Sandy Valley,Nevada,US,"https://www.google.com/maps/search/?api=1&query=Sandy Valley, Nevada, US"
12,Dome,944.0,4.94,268.0,Joshua Tree,California,US,"https://www.google.com/maps/search/?api=1&query=Joshua Tree, California, US"
13,Barn,417.0,4.95,111.0,Vero Beach,Florida,US,"https://www.google.com/maps/search/?api=1&query=Vero Beach, Florida, US"
14,Tiny home,372.0,4.97,174.0,Bolton,Vermont,US,"https://www.google.com/maps/search/?api=1&query=Bolton, Vermont, US"


In [ ]:
us_location_small = (
    df_location_us[
        df_location_us["country"] == " US"
    ]["price(in_dollar)","location", "state", "country"]
    .drop_duplicates()
    .reset_index(drop=True)
)

display(us_location_small.head(20))

NameError: name 'df_location_us' is not defined

In [ ]:
df_extracted["latitude"] = pd.Series(dtype="object")
df_extracted["longitude"] = pd.Series(dtype="object")

In [ ]:
display(df_extracted.head(30))

NameError: name 'df_extracted' is not defined

In [ ]:
print(df_extracted[["latitude", "longitude"]].dtypes)

latitude     object
longitude    object
dtype: object


In [ ]:
print(df_location_us.columns.tolist())

['price(in_dollar)', 'state', 'country', 'raiting', 'reviews', 'type', 'location', 'google_maps']


In [ ]:
!pip install folium
import folium
import pandas as pd
m = folium.Map(
    location=[39.8283, -98.5795],
    zoom_start=4
)

m


In [ ]:
import folium

for _, row in df_.iterrows():

    popup_text = f"""extrac
    <b>{row['type']}</b><br>
    Location: {row['location']}<br>
    State: {row['state']}<br>
    Price: ${row['price(in_dollar)']}<br>
    Rating: {row['raiting']}<br>
    Reviews: {row['reviews']}<br>
    <a href="{row['google_maps']}" target="_blank">
        Open in Google Maps
    </a>
    """

    folium.Marker(
        location=[row["latitude"], row["longitude"]],
        popup=folium.Popup(popup_text, max_width=1300)
    ).add_to(m)


m


NameError: name 'df_location_us' is not defined

In [ ]:
import pandas as pd
from geopy.geocoders import Nominatim
from IPython.display import displayimport numpy as np # Import numpy for np.nan

# Ensure 'location', 'state', and 'country' columns are strings for concatenation
# This also handles potential 'nan' float values by converting them to string 'nan'
df_location_us['location'] = df_location_us['location'].astype(str)
df_location_us['state'] = df_location_us['state'].astype(str)
df_location_us['country'] = df_location_us['country'].astype(str)

# Create or update the 'search_location' column for df_location_us
# Replacing 'nan' string with an empty string to avoid incorrect geocoding strings
df_location_us["search_location"] = (
    df_location_us["location"].replace('nan', '') + ", " +
    df_location_us["state"].replace('nan', '') + ", " +
    df_location_us["country"].replace('nan', '')
)

# Clean up any extra commas or spaces that might result from concatenation
df_location_us["search_location"] = df_location_us["search_location"].str.replace(r',\s*$', '', regex=True).str.strip()
df_location_us["search_location"] = df_location_us["search_location"].str.replace(r'^,\s*', '', regex=True).str.strip()

# Define the geocoding function (re-introducing it as it was deleted)
geolocator = Nominatim(user_agent="airbnb_analysis_colab_project", timeout=10)

def get_coordinates(place):
    if pd.isna(place) or str(place).strip() == '': # Handle NaN or empty strings
        return np.nan, np.nan
    try:
        location = geolocator.geocode(place, timeout=10)
        if location:
            return location.latitude, location.longitude
    except Exception as e:
        # Print a message for debugging, but don't stop the process
        print(f"Error geocoding '{place}': {e}")
    return np.nan, np.nan

# Apply the geocoding function to populate latitude and longitude
print("Geocoding locations for df_location_us...")
df_location_us[['latitude', 'longitude']] = df_location_us['search_location'].apply(
    lambda x: pd.Series(get_coordinates(x))
)

print("\nGeocoding complete. Displaying first 5 rows with new coordinates:")
display(df_location_us[['search_location', 'latitude', 'longitude']].head())

# Filter out rows where geocoding failed (latitude or longitude is NaN)
initial_rows = df_location_us.shape[0]
df_location_us.dropna(subset=['latitude', 'longitude'], inplace=True)
rows_after_geocoding_drop = df_location_us.shape[0]
print(f"Dropped {initial_rows - rows_after_geocoding_drop} rows due to failed geocoding.")

print("\nDataFrame info after geocoding and dropping null coordinates:")
df_location_us.info()